### ISSUE TITLE
Payment proof

### ISSUE THREAD

**Issue by isysd** [deginner/bitjws#6]
There are a number of scenarios where a request could include proof of payment. This proposal is meant to address a specific use case first: charging per request. An example use case might be charging $0.0001 per request to a public API server like a cryptocurrency explorer or data warehouse.

The changes to bitjws to accommodate this can be quite simple. Proposed is a new header `pay` with format `network:txid`. For example, it could be:

"bitcoin:a5610502963e5d49cd41d54a979da967c46990c8a19621d258bac2dcbbf20541"

or

"dash:90614410503de9098fb3b921f92a3c26d26984254ebf2e8a55a922b164e9998f"

This would be a transaction to the server's receiving address. The server will deduct the cost of the request from the balance created by the credit of the transaction.

For an example of accounting, a payment of 0.01 BTC could be paid upon registering for a service. Then a charge of 0.000001 would be deducted each request, yielding an initial credit of 10,000 requests.

Ideally the server would also respond nicely about payment and balance status, but that is a separate issue. Seems like a great case for the long dormant HTTP 402 code.

**Comment by welshjf** [comment]
Yeah, micropayments seem like an attractive way to monetize "public" APIs.

Why is this new header needed though? You'll have to authenticate to the service anyway -- otherwise anyone could spend your credits merely by knowing the txid -- so why not deduct from a credit balance associated with the account? Then you can fund it any way you like -- bitcoin, dash, cash sent by carrier pigeon -- and the client doesn't have to care.

**Comment by isysd** [comment]
Theoretically, the user could not register, and pay for each individual request. "Registering" i.e. another API call to set up some remote user record, is not required to authenticate or pay for a request.

The client probably will need to have a good amount of awareness of its spending, no matter what. You wouldn't want it to be too easy to trick the client into spending it's balance.

**Comment by welshjf** [comment]
Makes sense. For the second point though, I'm not sure what it means for a client to be aware of its spending. It can be written so as to economize its requests, but that doesn't require it to keep track of accounting details, which may be changing more frequently than a registration key. I guess the ideal would be to allow either usage.

**Comment by isysd** [comment]
# Git Guild Bounty Program

| Total | Dev               | Sales | Reporter | Tester(s) | Maintainer | Documenter |
|--------|------------------|---------|-------------|------------|-----------------|----------------|
|N/A    | assignee(s)   | N/A  | @isysd | @isysd, @d3brouille | @isysd  | @d3brouille |
| $2,750.00 | $1,650.00 | $137.50 | $137.50 | $550.00 | $137.50 | $137.50 |
